testplan/common/utils/table.py

```diff
@@ -32,58 +32,21 @@ def _std_table_from_list_of_lists(tbl):
         new_row = list(row)
         std_tbl.append(new_row)
         tbl_width = max(tbl_width, len(new_row))
-    else:  # empty list
-        return [[]]
     for row in std_tbl:
         pad_cells = tbl_width - len(row)
         row.extend([_TP_BLANK_CELL] * pad_cells)
+    if len(std_tbl) == 0:  # empty list
+        return [[]]
     return std_tbl
 
 
 def _std_table_from_list_of_dicts(tbl):
-    # type: (Sequence[Mapping[AnyStr, Any]]) -> List[List[Any]]
-    tbl_headers = []
-    tbl_headers_set = set()
-    tbl_dicts = list(tbl)
-    tbl_data = []
-    for maybe_val_dict in tbl_dicts:
-        # TODO: determine whether we should raise here, or whether we should
-        #       try harder to guess what is meant by a non-dict-like element
-        if not isinstance(maybe_val_dict, Mapping):
-            continue
-        val_dict = OrderedDict(maybe_val_dict)
-        for hdr, cell_val in six.iteritems(val_dict):
-            # this dance maintains the order of headers as much as possible
-            if hdr not in tbl_headers_set:
-                tbl_headers_set.add(hdr)
-                tbl_headers.append(hdr)
-            tgt_col_idx = tbl_headers.index(hdr)
-            # ensure there are at least `tgt_col_idx + 1` columns of something
-            # in `tbl_data`
-            checking_row_idx = 0
-            while checking_row_idx >= 0:
-                # ensure `tbl_data` contains at least `1 + checking_row_idx`
-                # rows
-                if len(tbl_data) <= checking_row_idx:
-                    # append an empty row
-                    tbl_data.append([])
-                curr_row = tbl_data[checking_row_idx]
-                # walk along `curr_row`, appending _TP_BLANK_CELL where needed,
-                # until `len(curr_row) >= (1 + tgt_col_idx)`
-                for col_idx in range(1 + tgt_col_idx):
-                    if len(curr_row) == col_idx:
-                        curr_row.append(_TP_BLANK_CELL)
-                # if
-                #   `curr_row[tgt_col_idx] == _TP_BLANK_CELL`
-                # then we can set
-                #   `curr_row[tgt_col_idx] = cell_val`
-                # else we'll need to check the same on the next row
-                if curr_row[tgt_col_idx] == _TP_BLANK_CELL:
-                    curr_row[tgt_col_idx] = cell_val
-                    checking_row_idx = -1
-                else:
-                    checking_row_idx += 1
-    return [tbl_headers] + tbl_data
+    dict_of_lists = {}
+    for dict_elem in tbl:
+        for hdr, val in six.iteritems(dict_elem):
```

Comment:
This does not sounds to be right when dictionaries in the list are non uniform (having different key sets).

Suggestion:
- compute the union of key sets from the dictionaries
- do whatever you do currently to add to the lists
- for the difference of the key set and the union, add None's or something that represent missing value

Or alternatively just reject those datasets to keep things simple, and get the user to provide nice data.

Reply:
The consolidate_columns method of TableEntry does the union of keys. Also it defines placeholder as "ABSENT" for empty cells. We only need one placeholder either "ABSENT" or "_TP_BLANK_CELL", need to consolidate.